### DISCUSSION TITLE
Issue with tsh ssh connection

### DISCUSSION THREAD

**Discussion by yzislin** [gravitational/teleport#13522] [Q&A]
Hello,
I am trying to ssh connect to nodes via tsh command. It fails with cert error.
I run :
tsh -d ssh --insecure --proxy=teleport.mydomain.com:443 --user=john

End of the error is:
`
DEBU [KEYSTORE] Returning Teleport TLS certificate "/root/.tsh/keys/teleport.mydomain.com/john-x509.pem" valid until "2022-06-15 07:23:08 +0000 UTC". client/keystore.go:307 DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:276 ERROR REPORT: Original Error: x509.UnknownAuthorityError x509: certificate signed by unknown authority Stack Trace: /go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:69 github.com/gravitational/teleport/lib/utils/proxy.directDial.dialALPNWithDeadline /go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:106 github.com/gravitational/teleport/lib/utils/proxy.directDial.Dial /go/src/github.com/gravitational/teleport/lib/client/api.go:2602 github.com/gravitational/teleport/lib/client.makeProxySSHClientWithTLSWrapper /go/src/github.com/gravitational/teleport/lib/client/api.go:2553 github.com/gravitational/teleport/lib/client.makeProxySSHClient /go/src/github.com/gravitational/teleport/lib/client/api.go:2525 github.com/gravitational/teleport/lib/client.(*TeleportClient).connectToProxy /go/src/github.com/gravitational/teleport/lib/client/api.go:2445 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToProxy.func1 /opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit User Message: x509: certificate signed by unknown authority
`
Thanks for your help.

**Comment by webvictim** [comment]
Does `tsh login --insecure --proxy=teleport.mydomain.com:443 --user=john` work?

What's the output of `tsh status`?

**Comment by yzislin** [comment]
WARNING: You are using insecure connection to SSH proxy https://teleport.mydomain.com
ERROR: WARNING:

  The proxy you are connecting to has presented a certificate signed by a
  unknown authority. This is most likely due to either being presented
  with a self-signed certificate or the certificate was truly signed by an
  authority not known to the client.

  If you know the certificate is self-signed and would like to ignore this
  error use the --insecure flag.

  If you have your own certificate authority that you would like to use to
  validate the certificate chain presented by the proxy, set the
  SSL_CERT_FILE and SSL_CERT_DIR environment variables respectively and try
  again.

  If you think something malicious may be occurring, contact your Teleport
  system administrator to resolve this issue.


root@myhostname:~# tsh status
ERROR: not logged in

**Reply by webvictim** [reply]
What’s in the `/etc/teleport.yaml` file on your Teleport server?

**Comment by yzislin** [comment]
`
version: v2
teleport:
  nodename: ip-123-456-789-123.ec2.internal
  data_dir: /opt/teleport
  log:
    output: /var/log/teleport/teleport.log
    severity: INFO
    format:
      output: text
  ca_pin: []
  diag_addr: ""
auth_service:
  enabled: "yes"
  listen_addr: 0.0.0.0:3025
  cluster_name: mycluster
  proxy_listener_mode: multiplex
  keep_alive_interval: 40s
  keep_alive_count_max: 3
ssh_service:
  enabled: "yes"
  labels:
    host: teleport
  commands:
  - name: hostname
    command: [hostname]
    period: 1m0s
proxy_service:
  enabled: "yes"
  web_listen_addr: 0.0.0.0:443
  public_addr: teleport.mydomain.com:443
  https_keypairs:
  - key_file: /opt/teleport/server.key
    cert_file: /opt/teleport/server.cert
  acme: {}
`

**Reply by webvictim** [reply]
How did you generate the `server.key` and `server.cert` files? It seems like the machine you’re using to run `tsh` doesn’t have the issuing CA available in its system trust store.

**Comment by yzislin** [comment]
I generated self-signed via openssl:
openssl req -x509 -newkey rsa:4096 -keyout key.pem -out cert.pem
Without password. I thought --insecure part should be okay.
My instance is an AWS EC2 amazon linux 2.

**Reply by webvictim** [reply]
Theoretically that should be ok. Does it work if you remove `https_keypairs` from the config? Teleport will generate its own self-signed key pair in that situation which may be more compatible.

**Comment by yzislin** [comment]
Just tried removed https_keypairs and restarted service. The same error happens when trying to "login" via tsh.

**Reply by webvictim** [reply]
Then I'm not sure what's going on, I'm afraid. I can't reproduce the issue you're having locally.

**Comment by yzislin** [comment]
After I comment out the key_pairs, do I need to wipe out any cached certs anywhere?
I just have a plain AWS linux 2 instance and teleport config file that you saw. Nothing fancy or unique.
Well, I do have application load balancer in front of teleport. It has a valid cert on it. 
Maybe this is the issue. Communication between ALB and Teleport.

Just tried to go to teleport directly (without generated certs) and it works just fine.

So ALB has publicly trusted valid cert. Does that mean the tsh client gets confused when it establishes connection to ALB and then receives self-signed stuff?

**Reply by webvictim** [reply]
If you're using an ALB in front of Teleport, the issue is entirely different. The problem is that the machine where you're running `tsh` doesn't trust the CA signing Amazon's certificates. If you're running on a Linux machine, I would probably look to install or update the `ca-certficates` package.

**Reply by yzislin** [reply]
So you are saying that Amazon's certificate authority is not trusted by default install of a linux?

**Reply by webvictim** [reply]
Without `ca-certificates` installed, no CAs will be trusted at all.

**Comment by yzislin** [comment]
Just tried to install it on a ubuntu box that I have been testing with and it is already installed.

**Comment by yzislin** [comment]
Unless I am wrong, it seems the original error complains about user cert (which is generated by teleport). This is weird.

**Reply by webvictim** [reply]
The original error is complaining that the certificate presented by the proxy `tsh` is connecting to isn't issued by a trusted authority.

What does `openssl s_client -connect teleport.mydomain.com:443 -showcerts | openssl x509 -noout -text` show?

You're definitely using a public ACM cert and not one from an Amazon private CA, right?

**Comment by yzislin** [comment]
`depth=2 C = US, O = Amazon, CN = Amazon Root CA 1
verify return:1
depth=1 C = US, O = Amazon, OU = Server CA 1B, CN = Amazon
verify return:1
depth=0 CN = teleport.mydomain.com
verify return:1
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number:
            xx:xx
        Signature Algorithm: sha256WithRSAEncryption
        Issuer: C = US, O = Amazon, OU = Server CA 1B, CN = Amazon
        Validity
            Not Before: Jun 10 00:00:00 2022 GMT
            Not After : Jul  9 23:59:59 2023 GMT
        Subject: CN = teleport.mydomain.com
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    xx:xx
                Exponent: xxxxx (0xxxxx)
        X509v3 extensions:
            X509v3 Authority Key Identifier: 
                keyid:xx:xx

            X509v3 Subject Key Identifier: 
                xx:xx
            X509v3 Subject Alternative Name: 
                DNS:teleport.mydomain.com
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 CRL Distribution Points: 

                Full Name:
                  URI:http://crl.sca1b.amazontrust.com/sca1b-1.crl

            X509v3 Certificate Policies: 
                Policy: 2.23.140.1.2.1

            Authority Information Access: 
                OCSP - URI:http://ocsp.sca1b.amazontrust.com
                CA Issuers - URI:http://crt.sca1b.amazontrust.com/sca1b.crt

            X509v3 Basic Constraints: critical
                CA:FALSE
            CT Precertificate SCTs: 
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : xx:xx
                    Timestamp : Jun 10 11:49:47.283 2022 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                xx:xx
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : xx:xx
                    Timestamp : Jun 10 11:49:47.341 2022 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                xx:xx
                Signed Certificate Timestamp:
                    Version   : v1 (0x0)
                    Log ID    : xx:xx
                    Timestamp : Jun 10 11:49:47.375 2022 GMT
                    Extensions: none
                    Signature : ecdsa-with-SHA256
                                xx:xx
    Signature Algorithm: sha256WithRSAEncryption
         xx:xx

read:errno=0
`

**Reply by webvictim** [reply]
Should be fine with that. What does `curl -v https://teleport.mydomain.com/webapi/ping` do?

**Comment by yzislin** [comment]
`*   Trying 123.123.123.123:443...
* TCP_NODELAY set
* Connected to teleport.mydomain.com (123.123.123.123) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*   CAfile: /etc/ssl/certs/ca-certificates.crt
  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* TLSv1.3 (IN), TLS handshake, Server hello (2):
* TLSv1.2 (IN), TLS handshake, Certificate (11):
* TLSv1.2 (IN), TLS handshake, Server key exchange (12):
* TLSv1.2 (IN), TLS handshake, Server finished (14):
* TLSv1.2 (OUT), TLS handshake, Client key exchange (16):
* TLSv1.2 (OUT), TLS change cipher, Change cipher spec (1):
* TLSv1.2 (OUT), TLS handshake, Finished (20):
* TLSv1.2 (IN), TLS handshake, Finished (20):
* SSL connection using TLSv1.2 / ECDHE-RSA-AES128-GCM-SHA256
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=teleport.mydomain.com
*  start date: Jun 10 00:00:00 2022 GMT
*  expire date: Jul  9 23:59:59 2023 GMT
*  subjectAltName: host "teleport.mydomain.com" matched cert's "teleport.mydomain.com"
*  issuer: C=US; O=Amazon; OU=Server CA 1B; CN=Amazon
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x564b325062f0)
> GET /webapi/ping HTTP/2
> Host: teleport.mydomain.com
> user-agent: curl/7.68.0
> accept: */*
> 
* Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
< HTTP/2 200 
< date: Mon, 20 Jun 2022 20:28:21 GMT
< content-type: application/json
< content-length: 427
< cache-control: no-cache, no-store, must-revalidate
< expires: 0
< pragma: no-cache
< 
* Connection #0 to host teleport.mydomain.com left intact
{"auth":{"type":"local","second_factor":"otp","preferred_local_mfa":"otp","has_motd":false},"proxy":{"kube":{"enabled":true,"listen_addr":"0.0.0.0:443"},"ssh":{"listen_addr":"0.0.0.0:443","tunnel_listen_addr":"0.0.0.0:443","public_addr":"teleport.mydomain.com:443"},"db":{"postgres_listen_addr":"0.0.0.0:443","mysql_listen_addr":"0.0.0.0:443"},"tls_routing_enabled":true},"server_version":"9.3.4","min_client_version":"8.0.0"}`

**Reply by webvictim** [reply]
OK. It looks like the HTTPS part is working fine and the certificate is trusted by the system.

One thing I've noticed - as you're using an AWS load balancer with an ACM cert, you'll need to disable TLS routing for Teleport to work correctly. To do this, edit your `/etc/teleport.yaml` on the server and perform the following actions:
- remove `version: v2`
- remove `proxy_listener_mode: multiplex`
- Restart Teleport

Make sure that your load balancer is forwarding at least ports 443, 3023 and 3024 to the server running Teleport.
- 443 and 3023 are required for `tsh login` and `tsh ssh`
- 443 and 3024 are required for agents/nodes to join via reverse tunnel

After you've done this, remove the `~/.tsh` directory and then try `tsh login` again. You shouldn't need `--insecure` at all, as the certificate is valid.

**Comment by yzislin** [comment]
I have removed these two lines from config. Allowed 3023 to the ALB and between ALB and Teleport.
tsh login is now throwing this error:
`ssh: handshake failed: ssh: overflow reading version string`

So is there a way to just setup all of the communications through proxy via 443 and without additional ports? it seems with v2 and multiplex everything works through 443. Well, minus the tsh login and tsh ssh.

Thanks for your help.

**Reply by webvictim** [reply]
Sadly it’s not technically possible to use both ACM and multiplexing on port 443. You would need to choose one or the other for anything other than the web UI to work.

**Comment by yzislin** [comment]
Just a thought. So my goal is to be able to provide users access to nodes via tsh ssh instead of web ui. I would like to keep version 2 config.
What if I utilize ssh_public_addr: setting in proxy service and point that to public IP of teleport server (ie bypass load balancer)?
I am just not sure how to configure that.

**Reply by webvictim** [reply]
You can set it up like this and then restart Teleport:

```yaml
proxy_service:
  public_addr: <LB hostname>:443
  ssh_public_addr: <server hostname>:3023
  tunnel_public_addr: <server hostname>:3024
```

This obviously won’t work for HA but will be fine for a single instance.

**Comment by yzislin** [comment]
So this means that I can TSH LOGIN and TSH SSH to server (ie teleport master) and specify nodes for ssh connection?
What would be the tsh login / tsh ssh connection string? I mean, what do I put for proxy?

So communication between teleport and nodes stays the same, right?

Thanks.

**Reply by webvictim** [reply]
`tsh login --proxy=teleport.mydomain.com:443` same as before. The other ports are automatically used by `tsh` behind the scenes.

Once you’re logged in, `tsh ls` shows nodes.
`tsh ssh user@hostname` will connect you.

Communication between Teleport and nodes will automatically use the new ports when you restart the node service after changing the proxy config.

**Comment by yzislin** [comment]
It is not working.
I've added two entries for ssh and tunnel to Teleport node with public IP of that box. I've restarted the service.
I am trying to connect via teleport.mydomain.com proxy. The same error as before, unknown authority for cert.

I've allowed 3023 and 3024 from client to teleport node.
443 is allowed to LB.

**Reply by webvictim** [reply]
Does that same `tsh login` command work on a different machine?

**Comment by yzislin** [comment]
I've tried on two different ones.
So looking at the debug log on the client that is initiating the connection, the problem arises right after user's cert gets downloaded.
And then:
`User Message: x509: certificate signed by unknown authority`
It feels like the issue is with client cert that comes through.

**Comment by yzislin** [comment]
So i've looked at the cert that gets downloaded after tsh login.
` X509v3 extensions:
            X509v3 Key Usage: critical
                Digital Signature, Key Encipherment
            X509v3 Extended Key Usage: 
                TLS Web Server Authentication, TLS Web Client Authentication
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Authority Key Identifier: 
                keyid:6C:33:5A:BA:F8:BB:75:D1:44:AD:E8:B3:EC:A8:D3:56:D1:20:9A:6`

Not sure if that `CA:FALSE` is right.
Also, the issuer is the cluster name.

**Comment by yzislin** [comment]
Oh, I could be wrong but I think after client gets their cert, it still tries to go to teleport.mydomain.com:443 for login.
`
INFO [KEYAGENT]  Loading SSH key for user "john" and cluster "mycluster". client/keyagent.go:202
INFO [CLIENT]    Connecting to proxy=teleport.mydomain.com:443 login="ubuntu" using TLS Routing client/api.go:2552
DEBU [KEYSTORE]  Returning Teleport TLS certificate "/root/.tsh/keys/teleport.mydomain.com/john-x509.pem" valid until "2022-06-22 06:37:54 +0000 UTC". client/keystore.go:307
DEBU [HTTP:PROX] No proxy set in environment, returning direct dialer. proxy/proxy.go:276

ERROR REPORT:
Original Error: x509.UnknownAuthorityError x509: certificate signed by unknown authority
Stack Trace:
	/go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:69 github.com/gravitational/teleport/lib/utils/proxy.directDial.dialALPNWithDeadline
	/go/src/github.com/gravitational/teleport/lib/utils/proxy/proxy.go:106 github.com/gravitational/teleport/lib/utils/proxy.directDial.Dial
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2602 github.com/gravitational/teleport/lib/client.makeProxySSHClientWithTLSWrapper
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2553 github.com/gravitational/teleport/lib/client.makeProxySSHClient
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2525 github.com/gravitational/teleport/lib/client.(*TeleportClient).connectToProxy
	/go/src/github.com/gravitational/teleport/lib/client/api.go:2445 github.com/gravitational/teleport/lib/client.(*TeleportClient).ConnectToProxy.func1
	/opt/go/src/runtime/asm_amd64.s:1581 runtime.goexit
User Message: x509: certificate signed by unknown authority

`

**Reply by webvictim** [reply]
I feel like this is probably a bug of some kind. I've asked the development team to take a look at this discussion to see whether there's anything obviously wrong.

**Comment by yzislin** [comment]
Any more ideas?

**Comment by yzislin** [comment]
Hi, Any updates on this?
It seems like a very common scenario with AWS.
Hopefully, you'll have a solution soon.


Thanks.
